### PR TITLE
feat: video component send events to kernel

### LINKED ISSUE #941: Get an event in the ECS when the video start/stop/pause
For static videos:
When a VideoTexture starts playing, there should be an event from Unity to Kernel signaling that the video started, the event should include the current position of the video in milliseconds.
When the video is paused or stopped, there should be also events informing the new state to the Kernel.
Once per second, if the video is playing, there should be a VideoProgressEvent(VideoTextureComponentId, offset, max) including the video component id and the milliseconds of the current playback. This event exists as a best effor to prevent synchronization/buffering issues between the Kernel and the Renderer.
Event definition
type VideoProgressEvent = {
  videoTextureId: string // component ID for the video texture
  status: number // 0 = paused, 1 = playing, 2 = buffering (paused)
  currentOffset: number // current absolute position of the video in milliseconds
  videoLength?: number // total video length in milliseconds, may be absent in streaming
}
For streaming:
When a VideoTexture starts streaming, there should be an event to know when the transmission started arriving. This event should include data about how long ago the transmission began for players that arrive late to the event (this information is typically included in any streaming data)

This is crucial for doing live events.

We often need to synchronize things that happen in the scene with a transmission, like light animations matching the music. The streaming always has a lag, which may often be up to 10 seconds, and depends on geography. If the instructions to run these animations are not properly synced to the actual arriving of the streaming, they will never be properly in sync with the video.

This task should do the changes relative to Unity and Kernel, but not the ecs
This task is relative to https://app.zenhub.com/workspaces/sdk-60febfe80dcd78000f01aca0/issues/decentraland/sdk/11

### PR DESCRIPTION
**Pull request by Kinerius**
## What does this PR change?

Fixes #941 

This PR containers a refactor to the WebVideoPlayer class that now receives a plugin interface so it could be mocked and fully tested.

## How to test the changes?

1. Go to: https://play.decentraland.zone/index.html?renderer=urn:decentraland:off-chain:renderer-artifacts:{branch_name}&{desired_url_params}
2. ...

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: unity-renderer/Assets/Scripts/MainScripts/DCL/WebInterface/Interface.cs
Comment (menduz) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/WebInterface/Interface.cs:1281]: Can this be replaced by the `ReportVideoProgressEvent`? Embedding all the data of the video in a single atomic message is better for the data-oriented approach of the ECS, it would add a component to the video that has a consistent full state all the times. What do you think?

The package is also missing the SceneId and the componentId of the video texture, that is necessary to redirect the event directly to the scene

**Review by menduz — CHANGES_REQUESTED** [review]

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/VideoTests.cs
Comment (menduz) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/VideoTests.cs:55]: Should the "status" value be an enum to prevent accidents?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (menduz) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:157]: I would pay attention to this method, probably calling `ReportVideoProgress` at the end at every change may be a good idea.

```suggestion
                if (model.playing)
                {
                    texturePlayer.Play();
                }
                else
                {
                    texturePlayer.Pause();
                }
                ReportVideoProgress();
```

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (menduz) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:223]: Same here, this number may better be a constant or enum somewhere to prevent accidents and to make the code easier to read and maintain

**Review by menduz — CHANGES_REQUESTED** [review]

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:157]: Yes you are right, it made sense last time when they were different

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:223]: Hmm this message structure is a DTO, I'm not sure if it's ok to create an enum for this because we already have a VideoState enum that has its own status
```c#
    public enum VideoState
    {
        NONE = 0,
        ERROR = 1,
        LOADING = 2,
        READY = 3,
        PLAYING = 4,
        BUFFERING = 5
    };
```
Maybe we can use the same status numbers at sdk.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:21]: Can we specify the measure of time for this? I guess is in seconds, so something like SECONDS_VIDEO_PROGRESS_UPDATE_INTERVAL

**Review by Suduck — APPROVED** [review]
LGTM! Small request!

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs]: I like this, made the change on last commit

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Plugins/IWebVideoPlayerPlugin.cs
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Plugins/IWebVideoPlayerPlugin.cs:21]:
```suggestion
        void Create(string id, string url, bool useHls);
        void Remove(string id);
        void TextureUpdate(string id, IntPtr texturePtr, bool isWebGL1);
        void Play(string id, float startTime);
        void Pause(string id);
        void SetVolume(string id, float volume);
        int GetHeight(string id);
        int GetWidth(string id);
        float GetTime(string id);
        float GetDuration(string id);
        int GetState(string id);
        string GetError(string id);
        void SetTime(string id, float second);
        void SetPlaybackRate(string id, float playbackRate);
        void SetLoop(string id, bool loop);
```
We should remove spaces before the semicolon.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Plugins/WebVideoPlayerNative.cs
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Plugins/WebVideoPlayerNative.cs:55]: Cool refactor. We can't move the static methods to a `Utils` class or something similar? This code would be far easier to follow with the change.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/WebVideoPlayerTests.cs
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/WebVideoPlayerTests.cs:8]: We are moving to the `Should...When` naming convention for tests. New test suites should abide to that convention if possible.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:201]: I think that this component logic is already getting pretty convoluted. We can't extract all the logic related to reporting the video progress to another class? (I don't mind if we have to create another coroutine for handling this).

**Review by BrianAmadori — COMMENTED** [review]
Pretty cool refactor, ty! Left some minor remarks to address before we merge it.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/VideoTests.cs
Comment (kuruk-mm) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/VideoTests.cs:121]:
```suggestion
                () => wasEventSent = true);
```

**Review by kuruk-mm — APPROVED** [review]

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Plugins/WebVideoPlayerNative.cs
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Plugins/WebVideoPlayerNative.cs:55]: Wouldn't the Utils class be the same as this then? you need to handle the existence of this DLL on multiple platforms and this class just does that

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/WebVideoPlayerTests.cs
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/Tests/WebVideoPlayerTests.cs]: Got it, will rename the class for now.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:201]: What about single level of abstraction? I prefer having less coroutines.
```c#
private IEnumerator OnUpdate()
{
    while (true)
    {
        UpdateDirtyState();
        UpdateVideoTexture();
        UpdateProgressReport();
        yield return null;
    }
}
```
Moving the report logic to another class could be another refactor or just leave it as technical debt

Comment (BrianAmadori) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Video/DCLVideoTexture.cs:201]: Sounds good!

**Review by BrianAmadori — APPROVED** [review]
Approving, will really appreciate if the feedback is addressed before the merge. TY!.

**Pull request merged**
